subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ErrorHandlingConfigurationResolver.java

```diff
@@ -60,7 +60,7 @@ public ErrorHandlingConfigurationResolver(ConfigurationResolver delegate) {
     public void resolveBuildDependencies(ConfigurationInternal configuration, ResolverResults results) {
         try {
             delegate.resolveBuildDependencies(configuration, results);
-        } catch (Throwable e) {
+        } catch (Exception e) {
```

Comment:
Can any checked exceptions be thrown here? Otherwise `catch RuntimeException` would be better. `catch Exception` can easily mask things like accidentally catching `InterruptedException`.

Reply:
I guess it's possible, since we compute build dependencies, we also compute a dependency graph, which could execute opaque rules.

Reply:
These rules can't throw checked exceptions though.

Reply:
why not? Groovy has no checked exception, nor Kotlin.

Reply:
Because it breaks any library that trusts the Java compiler's exception semantics.

Reply:
I can throw a checked exception from any Groovy or Kotlin code without having to declare anything, that's the difference. So the difference is that it doesn't force the enclosing method to declare a thrown exception. It means, in practice, that any Groovy, Kotlin (and I think Scala) code (or generated code) can throw any kind of exception, even if they implement an interface written in Java that doesn't expect checked exceptions. I would argue that's better since it doesn't force wrapping everything into a `RuntimeException`, or using "sneaky throwing", or makes the lives easier when writing lambda code.

Reply:
As I said, this breaks any Java code that expects (based on a long tradition) that only declared and runtime exceptions will be thrown. In fact many IDEs will give you a warning/error when you catch more than declared.

Reply:
That's not a good reason for me. The JVM allows it, we shouldn't bind our semantics to Java, I think. More so that we have Groovy and Kotlin code out there.

Reply:
It's actually even worse if we _don't_ do this and that user code in practice throws a checked exception, this would break our resolution engine and leak to the whole build like a fatal error. This shouldn't be the case.

Reply:
This PR is not about the retry logic. It's just about propagating Errors (like out of memory) instead of putting them into a `Collection<Throwable>` and continuing resolution.

Reply:
I agree smarter retry logic is out of scope. Probably worth it, but not as easy.